File: Data Structures & Algorithms/permutation-string/submission-3.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        // For my previous submission, I did CTRL + Enter
        // on accident and it submitted my unfinished code :(
        if(s1.length() > s2.length()){
            return false;
        }

        int right = 0;
        int left = 0;
        int[] freq1 = new int[26];
        int[] freq2 = new int[26];

        for(; right < s1.length(); right++){
            freq1[s1.charAt(right) - 'a']++;
            freq2[s2.charAt(right) - 'a']++;
        }
        
        for(; right < s2.length(); left++, right++){
            if(checkPermutation(freq1,freq2)){
                return true;
            }
            freq2[s2.charAt(left) - 'a']--;
            freq2[s2.charAt(right) - 'a']++;
        }
        return checkPermutation(freq1,freq2);
    }

    public boolean checkPermutation(int[] freq1, int[] freq2){
        for(int i = 0; i < 26; i++){
            if(freq1[i] != freq2[i]){
                return false;
            }
        }
        return true;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-3.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int maxArea = 0;

        for(int left = 0, right = heights.size() - 1; left < right;){
            if(heights[left] < heights[right]){
                maxArea = max(maxArea, heights[left] * (right - left));
                left++;
            }
            else{
                maxArea = max(maxArea, heights[right] * (right - left));
                right--;
            }
        }

        return maxArea;
    }
};
